activerecord/test/cases/scoping/named_scoping_test.rb

```diff
@@ -178,6 +178,30 @@ def test_first_and_last_should_not_use_query_when_results_are_loaded
     end
   end
 
+  def test_to_a_should_retain_preloaded_positions
+    classname = ActiveRecord::Base.connection.class.name[/[^:]*$/]
+    random_param = {
+      "SQLite3Adapter"    => "random()",
+      "Mysql2Adapter"     => "rand()",
+      "PostgreSQLAdapter" => "random()",
+      "OracleAdapter"     => "random()",
+      "FbAdapter"         => "random()"
+    }.fetch(classname) {
+      raise "need a random query parameter for #{classname}"
+    }
+
+    topics = Topic.order(random_param)
+    topics.stub :keep_preloaded_positions?, true do
```

Comment:
If this stub is necessary for testing, when will this case actually happen?

Reply:
We're using an explicit order-by-random to simulate the database returning differently-ordered results on two runs of a query that doesn't contain an ORDER BY clause. We couldn't think of a better way to reliably force that to happen, even though it's defined that it *might* happen.

It is pretty non-obvious, though. Would a comment help?

Reply:
At least In MySQL (InnoDB), it can expect a deterministic order even if the order is not explicitly given. This PR will break the following expectation in MySQL:

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "arel",  github: "rails/arel"
  gem "mysql2"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "mysql2", database: "test", user: "root")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :authors, force: true do |t|
    t.string :name, index: true
    t.timestamps
  end
end

class Author < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def setup
    @david = Author.create!(name: "David")
    Author.create!(name: "Mary")
    Author.create!(name: "Bob")
  end

  def test_30800
    authors = Author.select(:name)

    assert_equal @david.name, authors.first.name

    authors.create!(name: "Aaron")

    assert_equal ["Aaron", "Bob", "David", "Mary"], authors.map(&:name)
  end
end
```

So we should respect latest loaded records rather than partially loaded `@offsets` cache. #31006

Reply:
> At least In MySQL (InnoDB), it can expect a deterministic order even if the order is not explicitly given

Does MySQL *guarantee* that it's deterministic? That would seem to sacrifice possible future performance improvements. PostgreSQL, for example, will normally return the records in the order they appear on disk (which is deterministic), but if another session is already half way through a scan, it'll start from where-ever that one's up to -- so you might get the same sequence, but with a rotational offset.

Even assuming MySQL does make that [non-standard] guarantee, is it an implementation detail we wish to pass on? ISTM we'd still be within our rights, as an SQL abstraction layer, to use an undefined order to implement the best overall behaviour.

---

Your expectation leaves out a line:
```ruby
authors = Author.select(:name)
assert_equal "David", authors.first.name
authors.create!(name: "Aaron")
assert_equal "David", authors.first.name # <---
assert_equal ["Aaron", "Bob", "David", "Mary"], authors.map(&:name)
```

Does it still seem so obviously correct?

---

I'd prefer to make this pass:

```ruby
authors = Author.select(:name)
assert_equal "David", authors.first.name
authors.first.name = "Dawn"
assert_equal "Dawn", authors.first.name
authors.each { |a| nil }
assert_equal "Dawn", authors.first.name
```

Reply:
Since InnoDB always fetches data from either key, the result set is always a meaningful order.
So I'd prefer not to sort the result set by partially loaded offsets.
Another problem, `authors.to_a` is an natural order, but `authors.first` is ordered by `id ASC`, the order is not compatible.
So we should not reuse the offsets for latest loaded records.

Reply:
> Another problem, authors.to_a is an natural order, but authors.first is ordered by id ASC, the order is not compatible.

That's what we're trying to fix :smile:

I think the fact we currently sort on `id` for `first` already shows that we treat the results as unordered, regardless of what the database will do: if a user is expecting the result to be in a particular order, they must specify that with `.order`.

---

You bring up a good point, though: even without any data manipulation, we have this current behaviour on MySQL:

```ruby
authors = Author.select(:name)
assert_equal "David", authors.first.name
assert_equal "Bob", authors.to_a.first.name
assert_equal "Bob", authors.first.name
```

I really don't think that's sensible, predictable, or intuitive.

Reply:
I think that whether it is intuitive or not depends on whether the user knows the difference between `take` and `first`.

At least the following example is as I expected.

```ruby
authors = Author.select(:name)
assert_equal "Bob", authors.take.name
assert_equal "Bob", authors.to_a.take.name
assert_equal "Bob", authors.take.name
```

And also, since the order is incompatible, the cached offsets may not always be included in the result set.


```ruby
authors = Author.select(:name).limit(1)
assert_equal "David", authors.first.name
assert_equal 1, authors.to_a.size
```